src/models/DefaultPluginData.js

```diff
@@ -0,0 +1,185 @@
+  }
+
+  /**
+   * Get version of plugin core.
+   * @returns {String} Version of plugin core.
+   */
+  get coreVersion() {
+    return CORE_VERSION;
+  }
+
+  /**
+   * Get component by id.
+   * @param {String} id - Component id.
+   * @return {Component} Component or null.
+   */
+  getComponentById(id) {
+    return this.__getComponentById(this.components, id);
+  }
+
+  /**
+   * Get component by id.
+   * If components is not present in the provided components, look on their children.
+   * @param {Component[]} components - Components to search the id.
+   * @param {String} id - Id to search.
+   * @return {Component|null} Return searched component or null.
+   * @private
+   */
+  __getComponentById(components, id) {
+    // Remove sonar code smell, because its more optimized with a for loop.
+    for (let index = 0; index < components.length; index += 1) { // NOSONAR
+      if (components[index].id === id) {
+        return components[index];
+      }
+      const component = this.__getComponentById(components[index].children, id);
+      if (component) {
+        return component;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Create new component.
+   * @param {String} id - Component id.
+   * @param {ComponentDefinition} definition - Component definition.
+   */
+  addComponent(id, definition) {
+    this.components.push(new Component({
+      id,
+      name: id,
+      definition,
+    }));
+  }
+
+  /**
+   * Remove component by id, all attributes that used this component id
+   * and the links that refer to this component.
+   * @param {String} id - Component id.
+   */
+  removeComponentById(id) {
+    this.__removeComponentById(this.components, id);
+    this.__removeRefAttributeByValue(this.components, id);
+    this.links = this.links.filter(({ source, target }) => source !== id && target !== id);
+  }
+
+  /**
+   * Remove components by id.
+   * If not present in the provided components, look on their children.
+   * @param {Component[]} components - Components to search the id.
+   * @param {String} id - Id to remove.
+   * @return {boolean} Indicate if the component is removed.
+   * @private
+   */
+  __removeComponentById(components, id) {
+    // Remove sonar code smell, because its more optimized with a for loop.
+    for (let index = 0; index < components.length; index += 1) { // NOSONAR
+      if (components[index].id === id) {
+        components.splice(index, 1);
+        return true;
+      }
+      if (this.__removeComponentById(components[index].children, id)) {
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * Remove id in all attributes' value then if value is empty remove attribute.
+   * @param {Component[]} components - Components to search the id.
+   * @param {String} id - Id to remove.
+   * @private
+   */
+  __removeRefAttributeByValue(components, id) {
+    components.forEach((component) => {
+      if (component.children.length > 0) {
```

Comment:
This lambda should be a Component instance method, maybe ``Component.removeLinkToTarget(id)`` ?

Additionally, if component.children is null or undefined, this check will crash.

Reply:
children will never be null

Reply:
done